### ISSUE TITLE
“Transaction API error: Transaction not found”

### ISSUE THREAD

**Issue by heymartinadams** [prisma/prisma#16050]
### Bug description

**Context**: I’m trying to implement [Replicache](https://replicache.dev) using Prisma interactive transactions for an offline-first, instant UI. 🚀

**Bug**: While using transactions, I get a `Transaction API error: Transaction not found` error.

### How to reproduce

1. Clone https://github.com/ConsciousApps/replicache-demo
2. Obtain PlanetScale database test password via instructions in [README.md](https://github.com/ConsciousApps/replicache-demo#planetscale-test-database) file and replace placeholder in `.env` file
3. Install: `yarn` (or `npm install`)
4. Reset database (in case I, or someone else, has already populated it with data): `yarn db:reset`
5. Run `yarn dev`
6. Sign up or sign in
7. Once signed in, create 5 new to-dos in quick succession (creating new to-dos slowly does not produce this bug)

### Expected behavior

All transactions should be found.

### Prisma information

**Schema**: https://github.com/ConsciousApps/replicache-demo/blob/main/prisma/schema.prisma

**Prisma Transactions**:
https://github.com/ConsciousApps/replicache-demo/blob/main/pages/api/replicache/push.js#L24-L59
https://github.com/ConsciousApps/replicache-demo/blob/main/pages/api/replicache/pull.js#L20-L61

(bug shows up in either files)

### Environment & setup

- OS: macOS
- Database: PlanetScale MySQL
- Node.js version: `16.16.0`

### Prisma Version

```
4.6.0-dev.32 (latest)
```

**Comment by heymartinadams** [comment]
@janpio do you have any recommendation for an alternative MySQL database to use?

**Comment by heymartinadams** [comment]
Found what caused it. `Transaction not found` error went away once I removed this:

```js
} catch (err) {
  console.error(err)
  res.status(500).send(err.toString())
} finally {
  await prisma.$disconnect()
}
```

https://github.com/ConsciousApps/replicache-demo/commit/734f9ed3ece805aeb589e94689f4daf265b71563#diff-b5c776daef6284d72036715b72e4d5ff18cc0558dea8a0742f6c859e6f0f12b8L62-L68

**Comment by janpio** [comment]
We should try to figure out what is going on here.

**Comment by heymartinadams** [comment]
@janpio the [repo](https://github.com/ConsciousApps/replicache-demo) it’s a complete demo/reproduction so hopefully should be easy to set up and figure out.

**Comment by heymartinadams** [comment]
Hey @aqrln, I noticed you signed on yesterday... revisiting this repo today as well. Seems like there are a few bugs related to Replicache, I’ll see if I can work them out first and will let you know once I have.

**Comment by aqrln** [comment]
Hey @heymartinadams no worries, I already finished the investigation and was going to share an update with you tomorrow. The gist is the issue was triggered by calling `$disconnect` inside the `finally` clause of the try/catch block — i.e., every time after every single transaction regardless of whether succeeded or errored, which broke other concurrently running transactions as the disconnect actually happened. I was going to call this a bug and say that `$disconnect` should wait for the transactions to complete, but then I realised it has some implications in edge cases, so we will have to think about which behaviour should be considered correct together with the team.

I assume you actually meant to put the `$disconnect` into the `catch` clause and only put it into the `finally` clause by accident, but either way I’m curious what is the purpose of that? I would expect it to be redundant or even lead to performance issues due to re-establishing the database connection. Have you run into any issues where Prisma doesn’t properly recover from errors?

**Comment by heymartinadams** [comment]
Hi @aqrln, I didn’t know what I was doing regarding the `$disconnect`, so it’s more of a case of me having put this there without understanding it properly 🤦🏻‍♂

I updated the code quite a bit and am now able to actually get the caching system to work! But, I’m still running into “lag” issues, whereby if you mutate data quickly, it doesn’t update all mutations (see here for background).

https://user-images.githubusercontent.com/11673745/202359719-c68e612d-e687-4a66-936c-f5495bde16b9.mp4

And I’m wondering if this can be solved with Prisma’s Interactive Transactions? The issue here is that I need to access the database’s `serialized` isolation level for two different Prisma txs: one in `pull`, the other in `push`. Currently, when I’m pushing multiple mutations, Replicache does a “pull” at some point _during_ those txs using a reference value that hasn’t yet been updated, not until the end of the “push”. At least I think that’s what the issue is.

https://github.com/ConsciousApps/replicache-demo/blob/main/pages/api/replicache/push.js#L60-L64
https://github.com/ConsciousApps/replicache-demo/blob/main/pages/api/replicache/pull.js#L59-L63

**Comment by heymartinadams** [comment]
Hi @aqrln, I didn’t know what I was doing regarding the $disconnect, so it’s more of a case of me having put this there without understanding it properly 🤦🏻‍♂ Is the Prisma tx, the way it’s written without `$disconnect`, correct?

So I _finally_ figured out how to make it work, yay! 🎉 Looking forward to implementing it, and, once implemented, would love to connect with someone at Prisma to write up a blog post about it! Offline / undo-redo / and instant UI are definitely features in high demand and I think a blog post could add tremendous value to many devs. Please let me know!

**Comment by aqrln** [comment]
Hi @heymartinadams and sorry for the late reply!

> Looking forward to implementing it, and, once implemented, would love to connect with someone at Prisma to write up a blog post about it! Offline / undo-redo / and instant UI are definitely features in high demand and I think a blog post could add tremendous value to many devs. Please let me know!

Thank you, I passed your message to the team!

> Is the Prisma tx, the way it’s written without $disconnect, correct?

Yep, I didn't look into what happens inside the transaction itself in details, so I can't give much additional advice, but in the context of this issue just removing the $disconnect should be good! You can leave the try/catch block if you need it for any other reason, no need to remove it like in https://github.com/ConsciousApps/replicache-demo/commit/734f9ed3ece805aeb589e94689f4daf265b71563, just remove the `await prisma.$disconnect()` line.

In https://github.com/prisma/prisma-engines/pull/3391 I changed the error message from just `Transaction not found` to `Transaction not found. Transaction ID is invalid, refers to an old closed transaction Prisma doesn't have information about anymore, or was obtained before disconnecting.` Would that have been helpful in figuring out what happened and fixing the issue if the error message looked like that when you encountered it?

**Comment by aqrln** [comment]
I'm going to go ahead and close this issue since it has been resolved. Feel free to leave a comment or open a new issue if you think there is still something missing on our side.

We also have ideas for improvements of `$disconnect` behaviour to try to prevent mistakes like this — new GitHub issue(s) will be opened for that.

**Comment by heymartinadams** [comment]
Thank you @aqrln ✨ A more detailed error message might have been helpful at the time. Glad it’s made its way into `4.7.0`.

I’ll write up a blog post about Replicache shortly. https://github.com/simplerlist/replicache-demo

**Comment by aruns05** [comment]
@aqrln   Is any fix found for this ? I am facing same issue

**Comment by janpio** [comment]
This issue is closed, so we do not think there is a problem any more. If something is not working for you @aruns05, please open a new issue. Thanks.

**Comment by u007** [comment]
im still getting this for mongodb. it happen on a transaction block. using latest prisma 6.6.0
```

PrismaClientKnownRequestError:
Invalid `prisma.product.update()` invocation:


Transaction API error: Transaction not found. Transaction ID is invalid, refers to an old closed transaction Prisma doesn't have information about anymore, or was obtained before disconnecting.
       meta: [Object ...],
 clientVersion: "6.5.0",
       code: "P2028"

      at new PrismaClientKnownRequestError (/Users/james/www/ecomm/apps/backend/v2/node_modules/@prisma/client/runtime/library.js:21:1751)
      at handleRequestError (/Users/james/www/ecomm/apps/backend/v2/node_modules/@prisma/client/runtime/library.js:121:7534)
      at handleAndLogRequestError (/Users/james/www/ecomm/apps/backend/v2/node_modules/@prisma/client/runtime/library.js:121:6858)
      at <anonymous> (/Users/james/www/ecomm/apps/backend/v2/node_modules/@prisma/client/runtime/library.js:121:6565)
,
  name: "PrismaClientKnownRequestError",
}
```
